### PR TITLE
Allow the ability to specify a Kubernetes release version

### PR DESCRIPTION
**Pull request by onlydole**
Allow users to set the Kubernetes cluster version they want to use, otherwise default to the latest non-preview release available.

### CODE REVIEW

**Review by anubhavmishra — APPROVED** [review]

**Review by derkoe — APPROVED** [review]

**Review by derkoe — COMMENTED** [review]
Probably also add the version to the `default_node_pool`:

    orchestrator_version = var.kubernetes_version != null ? var.kubernetes_version : data.azurerm_kubernetes_service_versions.current.latest_version

Probably also use a different variable for this.

**Comment by onlydole** [comment]
Great point, @derkoe! I have updated my code and I ended up keeping the version the same for both the control plane and worker nodes for reliability's sake. 

Could I get your help in reviewing this, @yupwei68?

File: main.tf [hunk 1]
```diff
@@ -2,6 +2,11 @@ data "azurerm_resource_group" "main" {
   name = var.resource_group_name
 }
 
+data "azurerm_kubernetes_service_versions" "current" {
+  location        = data.azurerm_resource_group.main.location
+  include_preview = false
+}
+
```
Comment (yupwei68) [inline main.tf:9]: Since no version specified, `azurerm_kubernetes_cluster` will use the latest recommended version. Could we remove this to simplify the code?

File: main.tf [hunk 2]
```diff
@@ -12,6 +17,7 @@ resource "azurerm_kubernetes_cluster" "main" {
   location            = data.azurerm_resource_group.main.location
   resource_group_name = data.azurerm_resource_group.main.name
   dns_prefix          = var.prefix
+  kubernetes_version  = local.kubernetes_version
```
Comment (yupwei68) [inline main.tf:20]: kubernetes_version  = var.kubernetes_version

File: main.tf [hunk 3]
```diff
@@ -24,11 +30,12 @@ resource "azurerm_kubernetes_cluster" "main" {
   }
 
   default_node_pool {
-    name            = "nodepool"
-    node_count      = var.agents_count
-    vm_size         = var.agents_size
-    os_disk_size_gb = var.os_disk_size_gb
-    vnet_subnet_id  = var.vnet_subnet_id
+    orchestrator_version = local.kubernetes_version
```
Comment (yupwei68) [inline main.tf:33]: orchestrator_version = var.kubernetes_version

File: main.tf [hunk 4]
```diff
@@ -96,4 +103,7 @@ resource "azurerm_log_analytics_solution" "main" {
   }
 }
 
-
+locals {
+  # use the provided version, otherwise use the data source to find the latest Kubernetes version 
+  kubernetes_version = var.kubernetes_version != "" ? var.kubernetes_version : data.azurerm_kubernetes_service_versions.current.latest_version
+}
```
Comment (yupwei68) [inline main.tf:109]: not necessary?

File: variables.tf [hunk 5]
```diff
@@ -97,3 +97,9 @@ variable "sku_tier" {
   type        = string
   default     = "Free"
 }
+
+variable "kubernetes_version" {
+  description = "Specify which Kubernetes release to use. The default used is the latest Kubernetes version available in the region"
+  type        = string
+  default     = ""
```
Comment (yupwei68) [inline variables.tf:104]: default     = null

**Review by yupwei68 — COMMENTED** [review]
Thanks for your contribution. I've some suggestions to simplify the code by not using data source `azurerm_kubernetes_service_versions`. Since according to the definition, https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster

> kubernetes_version - (Optional) Version of Kubernetes specified when creating the AKS managed cluster. If not specified, the latest recommended version will be used at provisioning time (but won't auto-upgrade

BTW, would you add this feature into test/fixture to integrate it into test CI and add this feature into example in README?

File: main.tf (hunk 1 above)
Comment (onlydole) [reply main.tf:9]: Absolutely! Great point 👍🏼

**Comment by onlydole** [comment]
Thank you for the feedback, @yupwei68! I have made those changes. Please let me know if you need anything else done on that front, too!

File: README.md [hunk 6]
```diff
@@ -1,4 +1,5 @@
 # terraform-azurerm-aks
+
```
Comment (yupwei68) [inline README.md:2]: Please remove unnecessary changes.

File: README.md [hunk 7]
```diff
@@ -86,7 +88,7 @@ provider "kubernetes" {
 
 - [Configure Terraform for Azure](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/terraform-install-configure)
 
-We provide 2 ways to build, run, and test the module on a local development machine.  [Native (Mac/Linux)](#native-maclinux) or [Docker](#docker).
+We provide 2 ways to build, run, and test the module on a local development machine. [Native (Mac/Linux)](#native-maclinux) or [Docker](#docker).
```
Comment (yupwei68) [inline README.md:91]: I personally think one space is enough?

File: README.md [hunk 8]
```diff
@@ -133,7 +135,7 @@ $ rake full
 
 ### Docker
 
-We provide a Dockerfile to build a new image based `FROM` the `mcr.microsoft.com/terraform-test` Docker hub image which adds additional tools / packages specific for this module (see Custom Image section).  Alternatively use only the `microsoft/terraform-test` Docker hub image [by using these instructions](https://github.com/Azure/terraform-test).
+We provide a Dockerfile to build a new image based `FROM` the `mcr.microsoft.com/terraform-test` Docker hub image which adds additional tools / packages specific for this module (see Custom Image section). Alternatively use only the `microsoft/terraform-test` Docker hub image [by using these instructions](https://github.com/Azure/terraform-test).
```
Comment (yupwei68) [inline README.md:138]: Please remove unnecessary changes.

File: README.md [hunk 9]
```diff
@@ -165,7 +167,6 @@ This runs the full tests:
 $ docker run --rm azure-aks /bin/bash -c "bundle install && rake full"
 ```
 
-
```
Comment (yupwei68) [inline README.md:168]: here the same

File: README.md [hunk 10]
```diff
@@ -176,7 +177,7 @@ Originally created by [Damien Caro](http://github.com/dcaro) and [Malte Lantin](
 
 # Contributing
 
-This project welcomes contributions and suggestions.  Most contributions require you to agree to a
+This project welcomes contributions and suggestions. Most contributions require you to agree to a
```
Comment (yupwei68) [inline README.md:180]: here the same

File: main.tf [hunk 11]
```diff
@@ -24,22 +25,23 @@ resource "azurerm_kubernetes_cluster" "main" {
   }
 
   default_node_pool {
-    name            = "nodepool"
-    node_count      = var.agents_count
-    vm_size         = var.agents_size
-    os_disk_size_gb = var.os_disk_size_gb
-    vnet_subnet_id  = var.vnet_subnet_id
+    orchestrator_version = var.kubernetes_version
```
Comment (yupwei68) [inline main.tf:28]: I'd like to suggest add one more variable for `orchestrator_version`, because they're not necessary to be the same.

File: main.tf [hunk 12]
```diff
@@ -126,5 +128,3 @@ resource "azurerm_log_analytics_solution" "main" {
 
   tags = var.tags
 }
-
-
```
Comment (yupwei68) [inline main.tf:130]: here the same

File: variables.tf [hunk 13]
```diff
@@ -136,6 +136,9 @@ variable "rbac_aad_server_app_id" {
 
 variable "rbac_aad_server_app_secret" {
   description = "The Server Secret of an Azure Active Directory Application."
+}
```
Comment (yupwei68) [inline variables.tf:139]: Please don't remove the type and default for `rbac_aad_server_app_secret`

**Review by yupwei68 — COMMENTED** [review]
Thanks for your changes. 🚀 I've left some comments. Would you remove some unnecessary changes? Please run "terraform fmt" to keep the format. The fixture/test has failed for the change of `rbac_aad_server_app_secret`. Please run the following commands to ensure the changes have passed the test:
```
$ docker build --build-arg BUILD_ARM_SUBSCRIPTION_ID=$ARM_SUBSCRIPTION_ID --build-arg BUILD_ARM_CLIENT_ID=$ARM_CLIENT_ID --build-arg BUILD_ARM_CLIENT_SECRET=$ARM_CLIENT_SECRET --build-arg BUILD_ARM_TENANT_ID=$ARM_TENANT_ID -t azure-aks .
$ docker run --rm azure-aks /bin/bash -c "bundle install && rake full"
```

**Comment by onlydole** [comment]
Thank you for the feedback, @yupwei68. I have addressed the items you pointed out and tested locally, and everything is working as expected. 

Can you let me know if there are any more items you'd like addressed for this change?

**Review by yupwei68 — APPROVED** [review]
LGTM after changes.
